### PR TITLE
feat(sidecar): support jitter etc.

### PR DESCRIPTION
**Pull request by Stebalien**

### CODE REVIEW

File: pkg/sidecar/link.go [hunk 1]
```diff
@@ -96,48 +96,67 @@ func (l *NetlinkLink) init(idx uint16) error {
 }
 
 // Sets the bandwidth (in bytes per second).
-func (l *NetlinkLink) setBandwidth(idx uint16, rate uint64) error {
+func (l *NetlinkLink) setHtb(idx uint16, attrs netlink.HtbClassAttrs) error {
```
Comment (daviddias) [inline pkg/sidecar/link.go]: What does `Htb` stand for?

File: pkg/sidecar/link.go [hunk 2]
```diff
@@ -9,15 +9,15 @@ import (
...
 
 var (
 	rootHandle    = netlink.MakeHandle(1, 0)
 	defaultHandle = netlink.MakeHandle(1, 2)
 )
 
-var errMaxLatency = fmt.Errorf("latency can be at most %s", math.MaxUint32*time.Microsecond)
-
 // Link abstracts operations over a network interface.
 type NetlinkLink struct {
```
Comment (daviddias) [inline pkg/sidecar/link.go]: Why create a struct to subselect two properties of the netlink struct itself?

File: pkg/sidecar/link.go [hunk 3]
```diff
@@ -96,48 +96,67 @@ func (l *NetlinkLink) init(idx uint16) error {
...
-		netlink.NetemQdiscAttrs{
-			Latency: uint32(targetLatency),
-		},
+		attrs,
 	))
 }
 
-func (l *NetlinkLink) SetBandwidth(rate uint64) error {
-	return l.setBandwidth(0, rate)
+func toUs(t time.Duration) uint32 {
```
Comment (daviddias) [inline pkg/sidecar/link.go]: `us` is short for uInt32?

File: pkg/sidecar/link.go (hunk 2 above)
Comment (daviddias) [reply pkg/sidecar/link.go]: Gotcha, it is so that you can add methods to it after.

File: pkg/sidecar/link.go (hunk 1 above)
Comment (Stebalien) [reply pkg/sidecar/link.go]: The htb queuing class. https://linux.die.net/man/8/tc-htb

File: pkg/sidecar/link.go (hunk 3 above)
Comment (Stebalien) [reply pkg/sidecar/link.go]: Microsecond. I'll write that out.

File: pkg/sidecar/link.go (hunk 1 above)
Comment (Stebalien) [reply pkg/sidecar/link.go]: Ah, I forgot to update the docs. Thanks for catching that.

**Comment by Stebalien** [comment]
Ok, and this really does need tests. I never wired through some of the attributes.

**Comment by Stebalien** [comment]
But I'm not sure _how_ I'll test them without quite a bit of work...

File: pkg/sidecar/link.go [hunk 4]
```diff
@@ -95,59 +113,86 @@ func (l *NetlinkLink) init(idx uint16) error {
...
+func (l *NetlinkLink) setHtb(idx uint16, attrs netlink.HtbClassAttrs) error {
 	bandwidthHandle, _ := handlesForIndex(idx)
-	if rate == 0 {
-		rate = math.MaxUint64
-	}
 	return l.handle.ClassChange(netlink.NewHtbClass(
 		netlink.ClassAttrs{
 			LinkIndex: l.Attrs().Index,
 			Parent:    rootHandle,
 			Handle:    bandwidthHandle,
```
[the remark this answers is not in the document]
Comment (raulk) [reply pkg/sidecar/link.go]: Let’s reconcile these names?

File: pkg/sidecar/docker_instance.go [hunk 5]
```diff
@@ -303,11 +303,7 @@ func (dn *DockerNetwork) ConfigureNetwork(ctx context.Context, cfg *sync.Network
 		return fmt.Errorf("TODO: per-subnet bandwidth rules not supported")
 	}
 
-	// TODO: is 0 rate correct? How do I set "don't limit". Do I need to delete the rule?
-	if err := link.SetBandwidth(cfg.Default.Bandwidth); err != nil {
-		return err
-	}
-	if err := link.SetLatency(cfg.Default.Latency); err != nil {
+	if err := link.Shape(cfg.Default); err != nil {
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply pkg/sidecar/docker_instance.go]: The kernel just _rejects_ a zero bandwidth. Now, when the link-shape bandwidth is set to 0, I set it to int max (i.e., don't limit anything).

**Comment by Stebalien** [comment]
https://github.com/ipfs/testground/pull/222

**Pull request merged**